core/src/test/scala/unit/kafka/network/ConnectionQuotasTest.scala

```diff
@@ -0,0 +1,307 @@
+      ).get(5, TimeUnit.SECONDS)
+    } finally {
+      executor.shutdownNow()
+    }
+  }
+
+  @Test
+  def testNoConnectionLimitsByDefault(): Unit = {
+    val config = KafkaConfig.fromProps(brokerPropsWithDefaultConnectionLimits)
+    val connectionQuotas = new ConnectionQuotas(config, time)
+    addListenersAndVerify(config, connectionQuotas)
+
+    val executor = Executors.newFixedThreadPool(listeners.size)
+    try {
+      // verify there is no limit by accepting 10000 connections as fast as possible
+      val numConnections = 10000
+      val futures = listeners.values.map( listener =>
+        executor.submit((() => acceptConnections(connectionQuotas, listener, numConnections)): Runnable) )
+      futures.foreach(_.get(10, TimeUnit.SECONDS))
+      listeners.values.foreach { listener =>
+        assertEquals(s"${listener.listenerName.value()}",numConnections, connectionQuotas.get(listener.defaultClientIp))
+
+        // verify removing one connection
+        connectionQuotas.dec(listener.listenerName, listener.defaultClientIp)
+        assertEquals(s"Number of connections on $listener:", numConnections - 1, connectionQuotas.get(listener.defaultClientIp))
+
+        // calling dec() for an IP for which we didn't call inc() should throw an exception
+        intercept[IllegalArgumentException](connectionQuotas.dec(listener.listenerName, unknownHost))
+      }
+    } finally {
+      executor.shutdownNow()
+    }
+  }
+
+  @Test
+  def testMaxConnectionsPerIp(): Unit = {
+    val maxConnectionsPerIp = 17
+    val props = brokerPropsWithDefaultConnectionLimits
+    props.put(KafkaConfig.MaxConnectionsPerIpProp, maxConnectionsPerIp.toString)
+    val config = KafkaConfig.fromProps(props)
+    val connectionQuotas = new ConnectionQuotas(config, time)
+
+    addListenersAndVerify(config, connectionQuotas)
+
+    val executor = Executors.newFixedThreadPool(listeners.size)
+    try {
+      val externalListener = listeners("EXTERNAL")
+      executor.submit((() => acceptConnections(connectionQuotas, externalListener, maxConnectionsPerIp)): Runnable).get(5, TimeUnit.SECONDS)
+      assertEquals(s"Number of connections on $externalListener:", maxConnectionsPerIp, connectionQuotas.get(externalListener.defaultClientIp))
+
+      // all subsequent connections will be added to the counters, but inc() will throw TooManyConnectionsException for each of them
+      executor.submit((() => acceptConnectionsAboveIpLimit(connectionQuotas, externalListener, 2)): Runnable).get(5, TimeUnit.SECONDS)
+      assertEquals(s"Number of connections on $externalListener:", maxConnectionsPerIp + 2, connectionQuotas.get(externalListener.defaultClientIp))
+
+      // connections on the same listener but from a different IP should be accepted
+      executor.submit((() => acceptConnections(connectionQuotas, externalListener.listenerName, knownHost, maxConnectionsPerIp, 0)): Runnable).get(5, TimeUnit.SECONDS)
```

Comment:
A general comment. It would be great if we could break long lines in order to keep them under a reasonable size. I don't know if we have a strict guideline on this but we tend to do it.